tickit/adapters/interpreters/wrappers/splitting_interpreter.py

```diff
@@ -0,0 +1,114 @@
+
+        """
+        super().__init__()
+        self.interpreter: Interpreter[AnyStr] = interpreter
+        self.delimiter: AnyStr = message_delimiter
+        self.response_delimiter: AnyStr = response_delimiter
+
+    async def _handle_individual_messages(
+        self, adapter: Adapter, individual_messages: List[AnyStr]
+    ) -> List[Tuple[AsyncIterable[AnyStr], bool]]:
+        results = [
+            await self.interpreter.handle(adapter, message)
+            for message in individual_messages
+        ]
+        return results
+
+    async def _collect_responses(
+        self, results: List[Tuple[AsyncIterable[AnyStr], bool]]
+    ) -> Tuple[AsyncIterable[AnyStr], bool]:
+        """Combines results from handling multiple messages.
+
+        Takes the responses from when the wrapped interpreter handles multiple messages
+        and returns an appropriate composite repsonse and interrrupt. The composite
+        response is the concatentation of each of the individual responses, the
+        composite interrupt is a logical inclusive 'or' of all of the individual
+        responses.
+
+        Args:
+            results (List[Tuple[AsyncIterable[AnyStr], bool]]): a list of returned
+                values from the wrapped class' handle() method.
+
+        Returns:
+            Tuple[AsyncIterable[AnyStr], bool]:
+                A tuple of the asynchronous iterable of reply messages and a flag
+                indicating whether an interrupt should be raised by the adapter.
+        """
+        individual_responses, individual_interrupts = zip(*results)
+
+        response_list = [
+            response
+            for response_gen in individual_responses
+            async for response in response_gen
+        ]
+        response = self.response_delimiter.join(response_list)
+        resp = wrap_as_async_iterable(response)
+
+        interrrupt = any(individual_interrupts)
+
+        return resp, interrrupt
+
+    async def handle(
+        self, adapter: Adapter, message: AnyStr
+    ) -> Tuple[AsyncIterable[AnyStr], bool]:
+        """Splits a message and passes the resulting sub-messages on to an interpreter.
+
+        Splits a given message and passes the resulting sub-messages on to an
+        interpreter. The responses to the individual sub-messages are then combined
+        into a single response and returned.
+
+        Args:
+            adapter (Adapter): The adapter in which the function should be executed
+            message: (AnyStr): The message to be split up and handled.
+
+        Returns:
+            Tuple[AsyncIterable[Union[str, bytes]], bool]:
+                A tuple of the asynchronous iterable of reply messages and a flag
+                indicating whether an interrupt should be raised by the adapter.
+        """
+        # re.split(...) can contain empty strings and None - we discard these
+        individual_messages = [_ for _ in re.split(self.delimiter, message) if _]
```

Comment:
As well as `None`, re.split can also have additional empty strings in its return value, e.g.:
```
>>> re.split("pattern", "pattern at beginning")
['', ' at beginning'] 
```
If `""` is to be recognised as a command with a response, we don't want to introduce extra empty strings into the list of messages to be handled. I am currently relying on the falsyness of `""` to remove any extra empty strings introduced, whilst handling the case of completely empty message separately. Can we guarantee that an empty message sent to a device will never require a response?

Reply:
Good point! I hadn't considered the empty message case. However, it seems to me that only removing instances of `None` and foregoing re-adding an zero length message in empty case would achieve exactly what you're looking for with the added ability to send zero length messages in the middle of concatenated messages, e.g:
```python
    re.split(";", "hello;;world")
```
Giving:
```Python console
    ['hello', '', 'world']
```

And the zero length case:
```python
    re.split(";", "")
```
Giving:
```Python console
    ['']
```

What are your thoughts on this?

Reply:
I'm not sure this solves my problem (I may have misunderstood). I'll try to give an example:

Suppose you have a device that accepts 3 commands, `""`, `cmd1`, and `cmd2`, and needs to respond to each with `"0"`, `"1"`, and `"2"`, respectively. Suppose also that it is sent messages that look like `""` or `"/cmd1/cmd2"` etc.. I think wrapping a `CommandInterpreter` with a `SplittingInterpreter` using `message_delimiter="/"` (and, say, `response_delimiter=""`) makes sense here. If it is sent the message `"/cmd1/cmd2"` we would like it to pass on each of `["cmd1", "cmd2"]` to the wrapped interpreter, and eventually send back `"12"`; doing it your way would we instead be passing on `["", "cmd1, "cmd2"]` and sending back `"012"`?

I realise this is quite contrived, my 'real' example is related to https://github.com/dls-controls/tickit-devices/issues/10, but this has extra complications.

Reply:
In this case, where we wish to discard a leading delimiter we would need to wrap the `SplittingInterpreter` with a `BeheadingInterpreter` which chops off the first character. This still allows us to handle devices which receive a zero-length command at the beginning

Reply:
That does work... A slightly more complicated setup:

- The device accepts the commands `#`, `#x`, with x in {1,...8} and `P`
- The device is given messages consisting of several 'words' separated by `" "` 
- Each word is either a single command or two commands merged together, but only as `#xP` (not `PP` or `P#x` etc.)
- The number of responses must be equal to the number of words
- The device must respond to a empty message
- Say the response to `#x` is an empty message, the response to `P` is `"0"` and the response to `#` is `"1"`
- For each message there must be a reply message consisting of all individual responses followed by, say `"/"`

Some example messages, commands to be parsed and required replies:
- `"" -> [""] -> "/"`
- `"#1P #2 P" -> ["#1", "P", "#2", "P"] -> "0//0/"`
- `"P #2 #" -> ["P", "#2", "#"]  -> "0//1/"`

My approach to this was to wrap a `CommandInterpreter` with a `SplittingInterpreter` with `message_delimiter="(#[1-8]?)"` and `response_delimiter=""`, then wrap that in another `SplittingInterpreter` with `message_delimiter=" "` and `response_delimiter="/"`, but this only works with how I currently have the splitting set up. 

There's a reasonable chance I'm over-complicating things...

Reply:
Complex examples such as this should probably be handled with a more complex interpreter. For now, I think we should go with the `is not None` suggestion, even though it may potentially lead to unexpected behavior in some cases.